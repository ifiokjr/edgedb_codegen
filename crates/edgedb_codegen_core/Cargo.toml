[package]
name = "edgedb_codegen_core"
version = { workspace = true }
authors = { workspace = true }
categories = ["database"]
edition = { workspace = true }
homepage = { workspace = true }
keywords = ["edgedb", "database", "typesafe", "checked", "macros"]
license = { workspace = true }
readme = "readme.md"
repository = { workspace = true }
rust-version = { workspace = true }
description = "Utilities for codegen in the `edgedb_codegen` crate."

[dependencies]
check_keyword = { workspace = true }
edgedb-errors = { workspace = true }
edgedb-protocol = { workspace = true, features = ["all-types", "with-serde"] }
edgedb-tokio = { workspace = true, features = ["unstable", "derive"] }
heck = { workspace = true }
log = { workspace = true }
prettyplease = { workspace = true }
proc-macro2 = { workspace = true }
quote = { workspace = true }
syn = { workspace = true }
thiserror = { workspace = true }
tokio = { workspace = true, features = ["macros", "rt-multi-thread", "process"] }
typed-builder = { workspace = true }

[dev-dependencies]
assert2 = { workspace = true }
tokio = { workspace = true, features = ["time", "test-util"] }

[features]
with_bigint = []
with_bigdecimal = []
with_chrono = []
with_all = ["with_bigint", "with_bigdecimal", "with_chrono"]

[lints]
workspace = true
