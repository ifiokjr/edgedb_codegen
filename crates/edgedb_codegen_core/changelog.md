# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.2](https://github.com/ifiokjr/edgedb_codegen/compare/edgedb_codegen_core@v0.1.1...edgedb_codegen_core@v0.1.2) - 2024-08-27

### <!-- 0 -->🎉 Added
- *(edgedb_codegen_core)* add `prettyplease` format option

### <!-- 1 -->🐛 Bug Fixes
- increase minimum rust version

### <!-- 7 -->⚙️ Miscellaneous Tasks
- update coverage files
- support testing multiple rust versions

## [0.1.1](https://github.com/ifiokjr/edgedb_codegen/compare/edgedb_codegen_core@0.1.0...edgedb_codegen_core@0.1.1) - 2024-08-26

### <!-- 0 -->🎉 Added
- add support for `Range` types

### <!-- 7 -->⚙️ Miscellaneous Tasks
- improve changelog generation

## [0.1.0](https://github.com/ifiokjr/edgedb_codegen/releases/tag/edgedb_codegen_core-v0.1.0) - 2024-08-25

### 🎉 Added

- Initial release
