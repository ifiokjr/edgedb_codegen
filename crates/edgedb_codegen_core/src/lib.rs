use std::sync::Arc;

use check_keyword::CheckKeyword;
use edgedb_protocol::codec::CAL_DATE_DURATION;
use edgedb_protocol::codec::CAL_LOCAL_DATE;
use edgedb_protocol::codec::CAL_LOCAL_DATETIME;
use edgedb_protocol::codec::CAL_LOCAL_TIME;
use edgedb_protocol::codec::CAL_RELATIVE_DURATION;
use edgedb_protocol::codec::CFG_MEMORY;
use edgedb_protocol::codec::PGVECTOR_VECTOR;
use edgedb_protocol::codec::STD_BIGINT;
use edgedb_protocol::codec::STD_BOOL;
use edgedb_protocol::codec::STD_BYTES;
use edgedb_protocol::codec::STD_DATETIME;
use edgedb_protocol::codec::STD_DECIMAL;
use edgedb_protocol::codec::STD_DURATION;
use edgedb_protocol::codec::STD_FLOAT32;
use edgedb_protocol::codec::STD_FLOAT64;
use edgedb_protocol::codec::STD_INT16;
use edgedb_protocol::codec::STD_INT32;
use edgedb_protocol::codec::STD_INT64;
use edgedb_protocol::codec::STD_JSON;
use edgedb_protocol::codec::STD_STR;
use edgedb_protocol::codec::STD_UUID;
use edgedb_protocol::common::Capabilities;
use edgedb_protocol::common::Cardinality;
use edgedb_protocol::common::CompilationOptions;
use edgedb_protocol::common::IoFormat;
use edgedb_protocol::descriptors::Descriptor;
use edgedb_protocol::descriptors::ShapeElement;
use edgedb_protocol::descriptors::TupleElement;
use edgedb_protocol::descriptors::TypePos;
use edgedb_protocol::descriptors::Typedesc;
use edgedb_protocol::model::Uuid;
use edgedb_protocol::server_message::CommandDataDescription1;
use edgedb_tokio::create_client;
use edgedb_tokio::raw::Pool;
use edgedb_tokio::raw::PoolState;
use edgedb_tokio::Builder;
use heck::ToPascalCase;
use heck::ToSnakeCase;
use proc_macro2::TokenStream;
use quote::format_ident;
use quote::quote;
use syn::punctuated::Punctuated;
use syn::Token;
use tokio::runtime::Runtime;
use typed_builder::TypedBuilder;

pub use crate::constants::*;
pub use crate::errors::*;
pub use crate::utils::*;

mod constants;
mod errors;
mod utils;

/// Get the descriptor asynchronously.
pub async fn get_descriptor(query: &str) -> Result<CommandDataDescription1> {
	let builder = Builder::new().build_env().await?;
	let state = Arc::new(PoolState::default());
	let pool = Pool::new(&builder);
	let mut pool_connection = pool.acquire().await?;
	let connection = pool_connection.inner();
	let allow_capabilities = Capabilities::MODIFICATIONS | Capabilities::DDL;
	let flags = CompilationOptions {
		implicit_limit: None,
		implicit_typenames: false,
		implicit_typeids: false,
		explicit_objectids: true,
		allow_capabilities,
		io_format: IoFormat::Binary,
		expected_cardinality: Cardinality::Many,
	};

	Ok(connection.parse(&flags, query, &state).await?)
}

/// Get the descriptor synchronously.
pub fn get_descriptor_sync(query: &str) -> Result<CommandDataDescription1> {
	let rt = Runtime::new()?;
	let descriptor = rt.block_on(async { get_descriptor(query).await })?;

	Ok(descriptor)
}

pub fn generate_rust_from_query(
	descriptor: &CommandDataDescription1,
	name: &str,
	query: &str,
) -> Result<TokenStream> {
	let input_ident = format_ident!("{INPUT_NAME}");
	let output_ident = format_ident!("{OUTPUT_NAME}");
	let props_ident = format_ident!("{PROPS_NAME}");
	let query_ident = format_ident!("{QUERY_NAME}");
	let query_prop_ident = format_ident!("{QUERY_PROP_NAME}");
	let transaction_ident = format_ident!("{TRANSACTION_NAME}");
	let transaction_prop_ident = format_ident!("{TRANSACTION_PROP_NAME}");
	let module_name: syn::Ident = format_ident!("{}", name.to_snake_case());
	let input = descriptor.input.decode()?;
	let output = descriptor.output.decode()?;
	let mut tokens: TokenStream = TokenStream::new();

	explore_descriptor(
		ExploreDescriptorProps::builder()
			.typedesc(&input)
			.is_input()
			.is_root()
			.descriptor(input.root())
			.root_name(INPUT_NAME)
			.build(),
		&mut tokens,
	)?;
	explore_descriptor(
		ExploreDescriptorProps::builder()
			.typedesc(&output)
			.is_root()
			.descriptor(output.root())
			.root_name(OUTPUT_NAME)
			.build(),
		&mut tokens,
	)?;

	let query_method = match descriptor.result_cardinality {
		Cardinality::NoResult => quote!(execute),
		Cardinality::AtMostOne => quote!(query_single),
		Cardinality::One => quote!(query_required_single),
		Cardinality::Many | Cardinality::AtLeastOne => quote!(query),
	};

	let mut query_props = vec![quote!(#query_prop_ident: &#EXPORTS_IDENT::edgedb_tokio::Client)];
	let mut transaction_props =
		vec![quote!(#transaction_prop_ident: &mut #EXPORTS_IDENT::edgedb_tokio::Transaction)];
	let args = vec![
		quote!(#QUERY_CONSTANT),
		input.root().map_or(quote!(&()), |_| quote!(#props_ident)),
	];
	let inner_return = output.root().map_or(quote!(()), |_| quote!(#output_ident));
	let returns = wrap_token_with_cardinality(Some(descriptor.result_cardinality), inner_return);

	if input.root().is_some() {
		query_props.push(quote!(#props_ident: &#input_ident));
		transaction_props.push(quote!(#props_ident: &#input_ident));
	}

	let token_stream = quote! {
		pub mod #module_name {
			use ::edgedb_codegen::exports as #EXPORTS_IDENT;

			/// Execute the desired query.
			#[cfg(feature = "query")]
			pub async fn #query_ident(#(#query_props),*) -> core::result::Result<#returns, #EXPORTS_IDENT::edgedb_errors::Error> {
				#query_prop_ident.#query_method(#(#args),*).await
			}

			/// Compose the query as part of a larger transaction.
			#[cfg(feature = "query")]
			pub async fn #transaction_ident(#(#transaction_props),*) -> core::result::Result<#returns, #EXPORTS_IDENT::edgedb_errors::Error> {
				#transaction_prop_ident.#query_method(#(#args),*).await
			}

			#tokens

			/// The original query string provided to the macro. Can be reused in your codebase.
			pub const #QUERY_CONSTANT: &str = #query;
		}
	};

	Ok(token_stream)
}

fn wrap_token_with_cardinality(
	cardinality: Option<Cardinality>,
	token: TokenStream,
) -> TokenStream {
	let Some(cardinality) = cardinality else {
		return token;
	};

	match cardinality {
		Cardinality::NoResult | Cardinality::AtMostOne => quote!(Option<#token>),
		Cardinality::One => token,
		Cardinality::Many | Cardinality::AtLeastOne => quote!(Vec<#token>),
	}
}

#[derive(Debug, TypedBuilder)]
struct ExploreDescriptorProps<'a> {
	typedesc: &'a Typedesc,
	#[builder(setter(strip_bool(fallback = is_input_bool)))]
	is_input: bool,
	#[builder(setter(strip_bool(fallback = is_root_bool)))]
	is_root: bool,
	descriptor: Option<&'a Descriptor>,
	root_name: &'a str,
}

type PartialExploreDescriptorProps<'a> =
	ExploreDescriptorPropsBuilder<'a, ((&'a Typedesc,), (bool,), (bool,), (), ())>;

impl<'a> ExploreDescriptorProps<'a> {
	fn into_props(self) -> PartialExploreDescriptorProps<'a> {
		let Self {
			typedesc, is_input, ..
		} = self;

		Self::builder()
			.typedesc(typedesc)
			.is_input_bool(is_input)
			.is_root_bool(false)
	}
}

fn explore_descriptor(
	props @ ExploreDescriptorProps {
		typedesc,
		is_input,
		is_root,
		descriptor,
		root_name,
	}: ExploreDescriptorProps,
	tokens: &mut TokenStream,
) -> Result<Option<TokenStream>> {
	let root_ident = format_ident!("{root_name}");

	let Some(descriptor) = descriptor else {
		if is_root {
			tokens.extend(quote!(pub type #root_ident = ();));
		}

		return Ok(None);
	};

	match descriptor {
		Descriptor::Set(set) => {
			let set_descriptor = typedesc.get(set.type_pos).ok();
			let sub_root_name = format!("{root_name}Set");
			let props = props
				.into_props()
				.descriptor(set_descriptor)
				.root_name(&sub_root_name)
				.build();
			let result = explore_descriptor(props, tokens)?.map(|result| quote!(Vec<#result>));

			if is_root {
				tokens.extend(quote!(pub type #root_ident = #result;));
				Ok(Some(quote!(#root_ident)))
			} else {
				Ok(result)
			}
		}
		Descriptor::ObjectShape(object) => {
			let result = explore_object_shape_descriptor(
				StructElement::from_shape(&object.elements),
				typedesc,
				root_name,
				is_input,
				tokens,
			)?;

			Ok(result)
		}
		Descriptor::BaseScalar(base_scalar) => {
			let result = uuid_to_token_name(&base_scalar.id);

			if is_root {
				tokens.extend(quote!(pub type #root_ident = #result;));
				Ok(Some(quote!(#root_ident)))
			} else {
				Ok(Some(result))
			}
		}
		Descriptor::Scalar(scalar) => {
			let props = props
				.into_props()
				.descriptor(typedesc.get(scalar.base_type_pos).ok())
				.root_name(root_name)
				.build();

			explore_descriptor(props, tokens)
		}
		Descriptor::Tuple(tuple) => {
			let mut tuple_tokens = Punctuated::<_, Token![,]>::new();

			for (index, element) in tuple.element_types.iter().enumerate() {
				let sub_root_name = format!("{root_name}{index}");
				let result = explore_descriptor(
					ExploreDescriptorProps::builder()
						.typedesc(typedesc)
						.is_input_bool(is_input)
						.descriptor(typedesc.get(*element).ok())
						.root_name(&sub_root_name)
						.build(),
					tokens,
				)?;

				tuple_tokens.push(result);
			}

			let result = quote!((#tuple_tokens));

			if is_root {
				tokens.extend(quote!(pub type #root_ident = #result;));
				Ok(Some(quote!(#root_ident)))
			} else {
				Ok(Some(result))
			}
		}
		Descriptor::NamedTuple(named_tuple) => {
			let result = explore_object_shape_descriptor(
				StructElement::from_named_tuple(&named_tuple.elements),
				typedesc,
				root_name,
				is_input,
				tokens,
			)?;

			Ok(result)
		}
		Descriptor::Array(array) => {
			let array_descriptor = typedesc.get(array.type_pos).ok();
			let sub_root_name = format!("{root_name}Array");
			let props = props
				.into_props()
				.descriptor(array_descriptor)
				.root_name(&sub_root_name)
				.build();
			let result = explore_descriptor(props, tokens)?.map(|result| quote!(Vec<#result>));

			if is_root {
				tokens.extend(quote!(pub type #root_ident = #result;));
				Ok(Some(quote!(#root_ident)))
			} else {
				Ok(result)
			}
		}
		// TODO once `edgedb_protocol` is updated to 2.0 it should be possible to get the enum name.
		Descriptor::Enumeration(_) => {
			let result = Some(quote!(String));

			if is_root {
				tokens.extend(quote!(pub type #root_ident = #result;));
				Ok(Some(quote!(#root_ident)))
			} else {
				Ok(result)
			}
		}
		Descriptor::InputShape(object) => {
			let result = explore_object_shape_descriptor(
				StructElement::from_shape(&object.elements),
				typedesc,
				root_name,
				is_input,
				tokens,
			)?;

			Ok(result)
		}
		Descriptor::Range(range) => {
			let range_descriptor = typedesc.get(range.type_pos).ok();
			let sub_root_name = format!("{root_name}Range");
			let props = props
				.into_props()
				.descriptor(range_descriptor)
				.root_name(&sub_root_name)
				.build();
			let result = explore_descriptor(props, tokens)?
				.map(|result| quote!(#EXPORTS_IDENT::edgedb_protocol::model::Range<#result>));

			if is_root {
				tokens.extend(quote!(pub type #root_ident = #result;));
				Ok(Some(quote!(#root_ident)))
			} else {
				Ok(result)
			}
		}
		Descriptor::MultiRange(_) => todo!("`multirange` not in the `edgedb_protocol` crate"),
		Descriptor::TypeAnnotation(_) => todo!("type annotations are not supported"),
	}
}

pub fn explore_object_shape_descriptor(
	elements: Vec<StructElement<'_>>,
	typedesc: &Typedesc,
	root_name: &str,
	is_input: bool,
	tokens: &mut TokenStream,
) -> Result<Option<TokenStream>> {
	let mut impl_named_args = vec![];
	let mut struct_fields = vec![];
	let root_ident: syn::Ident = syn::parse_str(root_name)?;

	for element in elements {
		let descriptor = typedesc.get(element.type_pos()).ok();
		let name = &element.name();
		let safe_name = name.to_snake_case().into_safe();
		let safe_name_ident = format_ident!("{safe_name}");
		let pascal_name = name.to_pascal_case();
		let sub_root_name = format!("{root_name}{pascal_name}").into_safe();
		let sub_props = ExploreDescriptorProps::builder()
			.typedesc(typedesc)
			.is_input_bool(is_input)
			.descriptor(descriptor)
			.root_name(&sub_root_name)
			.build();
		let output = explore_descriptor(sub_props, tokens)?;
		let output_token = element.wrap(&output);
		let serde_annotation = (&safe_name != name).then_some(quote!(
			#[cfg_attr(feature = "serde", serde(rename = #name))]
		));
		let builder_fields = {
			match element.cardinality() {
				Cardinality::AtMostOne => {
					let fallback_ident = format_ident!("{safe_name_ident}_opt");
					Some(quote!(default, setter(into, strip_option(fallback = #fallback_ident))))
				}
				Cardinality::One => Some(quote!(setter(into))),
				Cardinality::Many => Some(quote!(default)),
				Cardinality::NoResult | Cardinality::AtLeastOne => None,
			}
		};
		let builder_annotation = (is_input && builder_fields.is_some()).then_some(quote!(
			#[cfg_attr(feature = "builder", builder(#builder_fields))]
		));

		struct_fields.push(quote! {
			#serde_annotation
			#builder_annotation
			pub #safe_name_ident: #output_token,
		});

		if is_input {
			impl_named_args.push(quote!(#name => self.#safe_name_ident.clone(),));
		}
	}

	let impl_tokens = is_input.then_some(quote! {
		impl #EXPORTS_IDENT::edgedb_protocol::query_arg::QueryArgs for #root_ident {
			fn encode(&self, encoder: &mut #EXPORTS_IDENT::edgedb_protocol::query_arg::Encoder) -> core::result::Result<(), #EXPORTS_IDENT::edgedb_errors::Error> {
				let map = #EXPORTS_IDENT::edgedb_protocol::named_args! {
					#(#impl_named_args)*
				};

				map.encode(encoder)
			}
		}
	});
	let typed_builder_tokens = is_input.then_some(
		quote!(#[cfg_attr(feature = "builder", derive(#EXPORTS_IDENT::typed_builder::TypedBuilder))]),
	);
	let struct_tokens = quote! {
		#[derive(Clone, Debug)]
		#typed_builder_tokens
		#[cfg_attr(feature = "query", derive(#EXPORTS_IDENT::edgedb_derive::Queryable))]
		#[cfg_attr(feature = "serde", derive(#EXPORTS_IDENT::serde::Serialize, #EXPORTS_IDENT::serde::Deserialize))]
		pub struct #root_ident {
			#(#struct_fields)*
		}

		#impl_tokens
	};

	tokens.extend(struct_tokens);

	Ok(Some(quote!(#root_ident)))
}

pub enum StructElement<'a> {
	Shape(&'a ShapeElement),
	Tuple(&'a TupleElement),
}

impl<'a> StructElement<'a> {
	pub fn from_shape(elements: &'a [ShapeElement]) -> Vec<StructElement<'a>> {
		elements.iter().map(From::from).collect::<Vec<_>>()
	}

	pub fn from_named_tuple(elements: &'a [TupleElement]) -> Vec<StructElement<'a>> {
		elements.iter().map(From::from).collect::<Vec<_>>()
	}

	pub fn name(&self) -> String {
		match self {
			StructElement::Shape(shape) => shape.name.clone(),
			StructElement::Tuple(tuple) => tuple.name.clone(),
		}
	}

	pub fn type_pos(&self) -> TypePos {
		match self {
			StructElement::Shape(shape) => shape.type_pos,
			StructElement::Tuple(tuple) => tuple.type_pos,
		}
	}

	pub fn wrap(&self, token: &Option<TokenStream>) -> TokenStream {
		if let Cardinality::AtMostOne = self.cardinality() {
			quote!(Option<#token>)
		} else {
			quote!(#token)
		}
	}

	pub fn cardinality(&self) -> Cardinality {
		match self {
			StructElement::Shape(shape) => shape.cardinality.unwrap_or(Cardinality::NoResult),
			StructElement::Tuple(_) => Cardinality::NoResult,
		}
	}
}

impl<'a> From<&'a ShapeElement> for StructElement<'a> {
	fn from(value: &'a ShapeElement) -> Self {
		StructElement::Shape(value)
	}
}

impl<'a> From<&'a TupleElement> for StructElement<'a> {
	fn from(value: &'a TupleElement) -> Self {
		StructElement::Tuple(value)
	}
}

fn uuid_to_token_name(uuid: &Uuid) -> TokenStream {
	match *uuid {
		STD_UUID => quote!(#EXPORTS_IDENT::uuid::Uuid),
		STD_STR => quote!(String),
		STD_BYTES => quote!(#EXPORTS_IDENT::bytes::Bytes),
		STD_INT16 => quote!(i16),
		STD_INT32 => quote!(i32),
		STD_INT64 => quote!(i64),
		STD_FLOAT32 => quote!(f32),
		STD_FLOAT64 => quote!(f64),
		#[cfg(not(feature = "with_bigdecimal"))]
		STD_DECIMAL => quote!(#EXPORTS_IDENT::edgedb_protocol::model::Decimal),
		#[cfg(feature = "with_bigdecimal")]
		STD_DECIMAL => quote!(#EXPORTS_IDENT::bigdecimal::BigDecimal),
		STD_BOOL => quote!(bool),
		#[cfg(not(feature = "with_chrono"))]
		STD_DATETIME => quote!(#EXPORTS_IDENT::edgedb_protocol::model::Datetime),
		#[cfg(feature = "with_chrono")]
		STD_DATETIME => quote!(#EXPORTS_IDENT::chrono::DateTime<#EXPORTS_IDENT::chrono::Utc>),
		#[cfg(not(feature = "with_chrono"))]
		CAL_LOCAL_DATETIME => quote!(#EXPORTS_IDENT::edgedb_protocol::model::LocalDatetime),
		#[cfg(feature = "with_chrono")]
		CAL_LOCAL_DATETIME => quote!(#EXPORTS_IDENT::chrono::NaiveDateTime),
		#[cfg(not(feature = "with_chrono"))]
		CAL_LOCAL_DATE => quote!(#EXPORTS_IDENT::edgedb_protocol::model::LocalDate),
		#[cfg(feature = "with_chrono")]
		CAL_LOCAL_DATE => quote!(#EXPORTS_IDENT::chrono::NaiveDate),
		#[cfg(not(feature = "with_chrono"))]
		CAL_LOCAL_TIME => quote!(#EXPORTS_IDENT::edgedb_protocol::model::LocalTime),
		#[cfg(feature = "with_chrono")]
		CAL_LOCAL_TIME => quote!(#EXPORTS_IDENT::chrono::NaiveTime),
		STD_DURATION => quote!(#EXPORTS_IDENT::edgedb_protocol::model::Duration),
		CAL_RELATIVE_DURATION => quote!(#EXPORTS_IDENT::edgedb_protocol::model::RelativeDuration),
		CAL_DATE_DURATION => quote!(#EXPORTS_IDENT::edgedb_protocol::model::DateDuration),
		STD_JSON => quote!(#EXPORTS_IDENT::edgedb_protocol::model::Json),
		#[cfg(not(feature = "with_bigint"))]
		STD_BIGINT => quote!(#EXPORTS_IDENT::edgedb_protocol::model::BigInt),
		#[cfg(feature = "with_bigint")]
		STD_BIGINT => quote!(#EXPORTS_IDENT::num_bigint::BigInt),
		CFG_MEMORY => quote!(#EXPORTS_IDENT::edgedb_protocol::model::ConfigMemory),
		PGVECTOR_VECTOR => quote!(#EXPORTS_IDENT::edgedb_protocol::model::Vector),
		_ => quote!(()),
	}
}

pub async fn get_types() -> Result<()> {
	let client = create_client().await?;
	let json = client.query_json(TYPES_QUERY, &()).await?;
	log::debug!("{}", json.as_ref());

	Ok(())
}
