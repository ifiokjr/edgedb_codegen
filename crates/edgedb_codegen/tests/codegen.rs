use std::path::PathBuf;
use std::thread;

use edgedb_codegen_core::*;
use proc_macro2::Span;
use rstest::fixture;
use rstest::rstest;

macro_rules! set_snapshot_suffix {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(format!($($expr,)*));
        let _guard = settings.bind_to_scope();
    }
}

#[fixture]
pub fn testname() -> String {
	thread::current()
		.name()
		.unwrap()
		.split("::")
		.last()
		.unwrap()
		.to_string()
}

#[rstest]
#[case::str("select 'i ❤️ edgedb'")]
#[case::bool("select true")]
#[case::int64("select 3")]
#[case::float64("select 314e-2")]
#[case::bigint("select 42n")]
#[case::decimal("select 42e+100n")]
#[case::uuid("select <uuid>'a5ea6360-75bd-4c20-b69c-8f317b0d2857'")]
#[case::datetime("select <datetime>'1999-03-31T15:17:00Z'")]
#[case::duration("select <duration>'45.6 seconds'")]
#[case::array("select [1, 2, 3]")]
#[case::tuple("select ('Apple', 7, true)")]
#[case::named_tuple("select (fruit := 'Apple', quantity := 3.14, fresh := true)")]
#[case::set(r#"select {"set", "of", "strings"}"#)]
#[case::empty_set("select <int64>{}")]
#[case::input_shape(
	"select { my_string := RelationshipType.Follow, my_number := 42, several_numbers := {1, 2, \
	 3}, array := [1, 2, 3] };"
)]
#[case::input_shape_with_args(r#"select { hello := "world", custom := <str>$custom }"#)]
#[case::object_shape("select Team {**}")]
#[case::object_shape_with_args(
	"select Team {**} filter .name like <str>$starts_with ++ '%' and .description like '%' ++ \
	 <str>$ends_with;"
)]
#[case::types_query(TYPES_QUERY)]
// #[case::range("select range(0, 10)")] // TODO: `Range` doesn't implement `Queryable` yet.
// #[case::bytes("select b'bina\\x01ry'")] // TODO: bytes don't implement `DecodeScalar` yet.
#[tokio::test]
#[rustversion::attr(not(nightly), ignore = "requires nightly")]
async fn codegen_literals(testname: String, #[case] query: &str) -> Result<()> {
	set_snapshot_suffix!("{}", testname);

	let relative_path = format!("tests/compile/codegen/{testname}.rs");
	let descriptor = get_descriptor(query).await?;
	let code = generate_rust_from_query(&descriptor, "example", query)?;
	let content = rustfmt(&code.to_string()).await?;

	// Check that the snapshot hasn't changed.
	insta::assert_snapshot!(&content);

	// Ensure that the produced rust is valid.
	prepare_compile_test(&content, &relative_path).await?;

	Ok(())
}

#[rstest]
#[case("insert_user")]
#[case("remove_user")]
#[tokio::test]
#[rustversion::attr(not(nightly), ignore = "requires nightly")]
async fn codegen_files(#[case] path: &str) -> Result<()> {
	set_snapshot_suffix!("{}", path);

	let query_path = resolve_path(format!("queries/{path}.edgeql"), Span::call_site())?;
	let relative_path = format!("tests/compile/codegen/{path}.rs");
	let query = tokio::fs::read_to_string(&query_path).await?;
	let descriptor = get_descriptor(&query).await?;
	let code = generate_rust_from_query(&descriptor, "example", &query)?;
	let content = rustfmt(&code.to_string()).await?;

	// Check that the snapshot hasn't changed.
	insta::assert_snapshot!(&content);

	// Ensure that the produced rust is valid.
	prepare_compile_test(&content, &relative_path).await?;

	Ok(())
}

const CRATE_DIR: &str = env!("CARGO_MANIFEST_DIR");

async fn prepare_compile_test(content: &str, relative_path: &str) -> Result<()> {
	let is_ci = std::env::var("CI")
		.ok()
		.is_some_and(|v| ["1", "true"].contains(&v.as_str()));

	let path = PathBuf::from(CRATE_DIR).join(relative_path);
	let generated = generate_contents(content).await?;

	let should_update = match tokio::fs::read_to_string(&path).await {
		Ok(current) => current != generated,
		Err(_) => true,
	};

	if should_update {
		assert2::assert!(!is_ci, "attempted updating compilation tests in CI");
		tokio::fs::write(&path, generated).await?;
	}

	Ok(())
}

async fn generate_contents(content: &str) -> Result<String> {
	let updated = format!("fn main() {{\n{content}\n}}");
	Ok(rustfmt(&updated).await?)
}
