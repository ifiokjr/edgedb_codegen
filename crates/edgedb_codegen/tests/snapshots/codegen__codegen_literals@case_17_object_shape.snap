---
source: crates/edgedb_codegen/tests/codegen.rs
expression: "&content"
---
pub mod example {
    use ::edgedb_codegen::exports as e;
    /// Execute the desired query.
    #[cfg(feature = "query")]
    pub async fn query(
        client: &e::edgedb_tokio::Client,
    ) -> core::result::Result<Vec<Output>, e::edgedb_errors::Error> {
        client.query(QUERY, &()).await
    }
    /// Compose the query as part of a larger transaction.
    #[cfg(feature = "query")]
    pub async fn transaction(
        conn: &mut e::edgedb_tokio::Transaction,
    ) -> core::result::Result<Vec<Output>, e::edgedb_errors::Error> {
        conn.query(QUERY, &()).await
    }
    pub type Input = ();
    #[derive(Clone, Debug)]
    #[cfg_attr(feature = "query", derive(e::edgedb_derive::Queryable))]
    #[cfg_attr(feature = "serde", derive(e::serde::Serialize, e::serde::Deserialize))]
    pub struct OutputWalletsSet {
        pub name: Option<String>,
        pub pubkey: String,
        pub created_at: e::chrono::DateTime<e::chrono::Utc>,
        pub id: e::uuid::Uuid,
        pub updated_at: e::chrono::DateTime<e::chrono::Utc>,
        pub primary: bool,
        pub description: Option<String>,
    }
    #[derive(Clone, Debug)]
    #[cfg_attr(feature = "query", derive(e::edgedb_derive::Queryable))]
    #[cfg_attr(feature = "serde", derive(e::serde::Serialize, e::serde::Deserialize))]
    pub struct Output {
        pub slug: String,
        pub id: e::uuid::Uuid,
        pub created_at: e::chrono::DateTime<e::chrono::Utc>,
        pub updated_at: e::chrono::DateTime<e::chrono::Utc>,
        pub description: Option<String>,
        pub name: String,
        pub wallets: Vec<OutputWalletsSet>,
    }
    /// The original query string provided to the macro. Can be reused in your codebase.
    pub const QUERY: &str = "select Team {**}";
}
